# Snakefile
# Load the configuration file
configfile: "config.yaml"

import csv

params = {}
uids = []

with open(config["params_file"]) as f:
    reader = csv.DictReader(f)
    for row in reader:
        uid = row["uid"]
        params[uid] = row
        uids.append(uid)


OUTPUT_DIR = config["output_dir"]
THREADS = config.get("threads", 4)

# ---- rule all ----
rule all:
    input:
        expand(f"{OUTPUT_DIR}/{{uid}}.tsinfer.trees", uid=uids)


# ---- VCF to Zarr ----
rule convert_vcf_to_zarr:
    input:
        vcf=lambda wc: params[wc.uid]["vcf_file"]
    output:
        zarr_dir=directory(f"{OUTPUT_DIR}/{{uid}}.vcz")
    params:
        outbase=lambda wc: f"{OUTPUT_DIR}/{wc.uid}"
    container: config.get("singularity")
    shell:
        """
        vcf2zarr explode {input.vcf} {params.outbase}.icf
        vcf2zarr encode {params.outbase}.icf {output.zarr_dir}
        """

# ---- tsinfer ----
rule run_tsinfer:
    input:
        zarr_dir=f"{OUTPUT_DIR}/{{uid}}.vcz"
    output:
        trees=f"{OUTPUT_DIR}/{{uid}}.tsinfer.trees"
    params:
        mu=lambda wc: params[wc.uid]["mu"],
        contig=lambda wc: params[wc.uid]["contig"],
        start=lambda wc: params[wc.uid]["start"],
        end=lambda wc: params[wc.uid]["end"],
        recomb_map=lambda wc: params[wc.uid]["recomb_map"],
        fasta=lambda wc: params[wc.uid]["ancestral_fasta"],
        #mask=lambda wc: params[wc.uid]get("mask_bed", ""), #FIXME
        seed=lambda wc: params[wc.uid]["seed"],
        script_path = "./scripts/run_tsinfer.py"
    container: config.get("singularity")
    threads: THREADS
    shell:
        """
        python {params.script_path} \
            --zarr {input.zarr_dir} \
            --fasta {params.fasta} \
            --chrom {params.contig} \
            --recomb-map {params.recomb_map} \
            --mut-rate {params.mu} \
            --outdir {OUTPUT_DIR} \
            --threads {threads}
        """
